### PR TITLE
fix Access restriction to form answers

### LINKED ISSUE #3179: users are able to view all forms by simply changing the ID in the URL
Dear Support Team,

I am writing to report a bug that I have found in the Formcreator plugin, version 2.13.4. When using this plugin in GLPI version 10.0.6, I have discovered that users are able to view all forms by simply changing the ID in the URL.

This security issue is concerning as it allows unauthorized access to sensitive information. I have tried to reproduce the issue multiple times and can confirm that it is present consistently.

Please let me know if there is a workaround or a fix available for this issue. If any additional information is required, please let me know and I will be happy to provide it.

Thank you for your attention to this matter.

### PR DESCRIPTION
**Pull request by btry**
fix #3179

### CODE REVIEW

File: inc/formanswer.class.php [hunk 1]
```diff
@@ -127,8 +127,10 @@ public function canViewItem() {
 
       $groupUser = new Group_User();
       $groups = $groupUser->getUserGroups($currentUser);
-      if (in_array($this->fields['users_id_validator'], $groups)) {
-         return true;
+      foreach ($groups as $group) {
+         if ($this->fields['groups_id_validator'] == $group['id']) {
+            return true;
+         }
```
Comment (btry) [inline inc/formanswer.class.php:133]: bad implementation fixed : $groups was not an array of group IDs. It actually contains an array of array of group fields.

File: inc/formanswer.class.php [hunk 2]
```diff
@@ -170,32 +172,36 @@ public function canViewItem() {
       $ticket_user_table = Ticket_User::getTable();
       $item_ticket_table = Item_Ticket::getTable();
       $request = [
-         'SELECT' => Ticket_User::getTableField(User::getForeignKeyField()),
+         'SELECT' => [
+            Ticket_User::getTableField(User::getForeignKeyField()),
+            Ticket::getTableField('id'),
```
Comment (btry) [inline inc/formanswer.class.php:177]: Ticket ID added for debug needs, not a big issue to keep it in the SELECT clause

File: inc/formanswer.class.php [hunk 3]
```diff
@@ -170,32 +172,36 @@ public function canViewItem() {
...
                ],
             ],
             $item_ticket_table => [
                'FKEY' => [
                   $item_ticket_table => 'tickets_id',
                   $ticket_table => 'id',
                   ['AND' => [
-                     Item_Ticket::getTableField('itemtype') =>  self::getType(),
+                     Item_Ticket::getTableField('itemtype') => self::getType(),
+                     Item_Ticket::getTableField('items_id') => $this->getID(),
```
Comment (btry) [inline inc/formanswer.class.php:196]: Results should be filtered by ID of form answer being checked

File: inc/formanswer.class.php [hunk 4]
```diff
@@ -170,32 +172,36 @@ public function canViewItem() {
...
-
          ]
       ];
 
-      foreach ($DB->request($request) as $row) {
-         if ($row[User::getForeignKeyField()] == $currentUser) {
-            return true;
-         }
+      if ($DB->request($request)->count() > 0) {
+         return true;
```
Comment (btry) [inline inc/formanswer.class.php:204]: now, if there is at least 1 row found, we know that the user is an actor of a ticket linked to this form answer

File: tests/3-unit/PluginFormcreatorForm_Validator.php [hunk 5]
```diff
@@ -30,6 +30,11 @@
  */
 namespace tests\units;
 use GlpiPlugin\Formcreator\Tests\CommonTestCase;
+use Group;
+use User;
+use Group_User;
+use PluginFormcreatorForm;
+use PluginFormcreatorLinker;
```
Comment (btry) [inline tests/3-unit/PluginFormcreatorForm_Validator.php:37]: remove canonical class names; extra refactor in this file

**Review by AdrienClairembault — APPROVED** [review]

**Pull request merged**